pandas/tests/indexes/datetimes/test_timezones.py

```diff
@@ -0,0 +1,1047 @@
+        dr = date_range(datetime(2011, 6, 1, 0), periods=10,
+                        freq=pd.offsets.Hour())
+        localized = dr.tz_localize(tz)
+        localized_infer = dr.tz_localize(tz, ambiguous='infer')
+        tm.assert_index_equal(localized, localized_infer)
+        with tm.assert_produces_warning(FutureWarning):
+            localized_infer_old = dr.tz_localize(tz, infer_dst=True)
+        tm.assert_index_equal(localized, localized_infer_old)
+
+    @pytest.mark.parametrize('tz', [pytz.timezone('US/Eastern'),
+                                    gettz('US/Eastern')])
+    def test_dti_tz_localize_ambiguous_times(self, tz):
+        # March 13, 2011, spring forward, skip from 2 AM to 3 AM
+        dr = date_range(datetime(2011, 3, 13, 1, 30), periods=3,
+                        freq=pd.offsets.Hour())
+        with pytest.raises(pytz.NonExistentTimeError):
+            dr.tz_localize(tz)
+
+        # after dst transition, it works
+        dr = date_range(datetime(2011, 3, 13, 3, 30), periods=3,
+                        freq=pd.offsets.Hour(), tz=tz)
+
+        # November 6, 2011, fall back, repeat 2 AM hour
+        dr = date_range(datetime(2011, 11, 6, 1, 30), periods=3,
+                        freq=pd.offsets.Hour())
+        with pytest.raises(pytz.AmbiguousTimeError):
+            dr.tz_localize(tz)
+
+        # UTC is OK
+        dr = date_range(datetime(2011, 3, 13), periods=48,
+                        freq=pd.offsets.Minute(30), tz=pytz.utc)
+
+    @pytest.mark.parametrize('tzstr', ['US/Eastern', 'dateutil/US/Eastern'])
+    def test_dti_tz_localize_pass_dates_to_utc(self, tzstr):
+        strdates = ['1/1/2012', '3/1/2012', '4/1/2012']
+
+        idx = DatetimeIndex(strdates)
+        conv = idx.tz_localize(tzstr)
+
+        fromdates = DatetimeIndex(strdates, tz=tzstr)
+
+        assert conv.tz == fromdates.tz
+        tm.assert_numpy_array_equal(conv.values, fromdates.values)
+
+    @pytest.mark.parametrize('prefix', ['', 'dateutil/'])
+    def test_dti_tz_localize(self, prefix):
+        tzstr = prefix + 'US/Eastern'
+        dti = DatetimeIndex(start='1/1/2005', end='1/1/2005 0:00:30.256',
+                            freq='L')
+        dti2 = dti.tz_localize(tzstr)
+
+        dti_utc = DatetimeIndex(start='1/1/2005 05:00',
+                                end='1/1/2005 5:00:30.256', freq='L', tz='utc')
+
+        tm.assert_numpy_array_equal(dti2.values, dti_utc.values)
+
+        dti3 = dti2.tz_convert(prefix + 'US/Pacific')
+        tm.assert_numpy_array_equal(dti3.values, dti_utc.values)
+
+        dti = DatetimeIndex(start='11/6/2011 1:59', end='11/6/2011 2:00',
+                            freq='L')
+        with pytest.raises(pytz.AmbiguousTimeError):
+            dti.tz_localize(tzstr)
+
+        dti = DatetimeIndex(start='3/13/2011 1:59', end='3/13/2011 2:00',
+                            freq='L')
+        with pytest.raises(pytz.NonExistentTimeError):
+            dti.tz_localize(tzstr)
+
```

Comment:
prob should dedupe these tz parameters into some nicely named fixtures (and push up what you can to conftest)

Reply:
next PR